styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -0,0 +1,342 @@
+     * have already been initialized and incremented by another process.
+     */
+    private static void initialize(Datastore datastore, String counterId) {
+      for (int i = 0; i < NUM_SHARDS; i++) {
+        final int shardIndex = i;
+        final Key shardKey = datastore.newKeyFactory().setKind(KIND_COUNTER_SHARD)
+            .newKey(counterId + "-" + shardIndex);
+        datastore.runInTransaction(transaction -> {
+          final Entity shard = transaction.get(shardKey);
+          if (shard == null) {
+            transaction.put(Entity.newBuilder(shardKey)
+                                .set(PROPERTY_COUNTER_ID, counterId)
+                                .set(PROPERTY_SHARD_INDEX, shardIndex)
+                                .set(PROPERTY_SHARD_VALUE, 0)
+                                .build());
+          }
+          return null;
+        });
+      }
+    }
+
+    private static Map<Integer, Long> fetchShards(Datastore datastore, String counterId) {
+      final EntityQuery queryShards = EntityQuery.newEntityQueryBuilder()
+          .setKind(KIND_COUNTER_SHARD)
+          .setFilter(PropertyFilter.eq(PROPERTY_COUNTER_ID, counterId))
+          .setLimit(NUM_SHARDS)
+          .build();
+      final QueryResults<Entity> shards = datastore.run(queryShards);
+      final Map<Integer, Long> fetchedShards = new HashMap<>();
+      while (shards.hasNext()) {
+        Entity shard = shards.next();
+        fetchedShards
+            .put((int) shard.getLong(PROPERTY_SHARD_INDEX), shard.getLong(PROPERTY_SHARD_VALUE));
+      }
+      return fetchedShards;
+    }
+
+    private boolean isRecent() {
+      return updatedAt.plus(CACHE_EXPIRY_DURATION.toMillis(), ChronoUnit.MILLIS)
+          .isAfter(time.get());
+    }
+
+    /**
+     * The shard's capacity is calculated as:
+     * 1/NUM_SHARDS part of the total count capacity
+     * plus an extra 1 unit of the remainder (whenever limit % NUM_SHARDS > 0) depending on the shardIndex
+     * The first (limit % NUM_SHARDS) shards get 1 unit extra, and the rest get 0.
+     *
+     * </p>ex. If limit=5 for a given counter and NUM_SHARDS=3,
+     * then the distribution of capacity between the 3 shards will be [2, 2, 1]
+     */
+    private long shardCapacity(int shardIndex) {
+      return limit / NUM_SHARDS + (shardIndex < limit % NUM_SHARDS ? 1 : 0);
+    }
+
+    /**
+     * Returns shard index which _likely_ could be successfully updated by delta, according to our
+     * cached view of the state in Datastore.
+     */
+    private int pickShardWithSpareCapacity(long delta) {
+      List<Integer> candidates = new ArrayList<>();
+
+      for (int i : shards.keySet()) {
+        if (shards.containsKey(i) && Range
+            .closed(0L, shardCapacity(i)).contains(shards.get(i) + delta)) {
+          candidates.add(i);
+        } else if (!shards.containsKey(i)) {
+          LOG.warn("Shard {} for counter {} is not present in local cache", i, counterId);
```

Comment:
This can happen in normal operation due to eventual consistency. So I think we shouldn't warn about that, better avoid warning fatigue. If the counter doesn't get out of this state for a prolonged period, _then_ we or Datastore have a bug or incident, that would be worth warning about, but detecting that condition requires more code and investment, and as such should better be decoupled from the counter algorithm implementation...